src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1186,132 @@ void Client::HandleTimer(void)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+
+    const otLinkedBuffer *data;
+    size_t                totalReceived = 0;
+    Message              *message       = nullptr;
+    Response              response;
+    QueryType             type;
+    Error                 responseError;
+    uint16_t              length = 0;
+
+    if (aEndOfStream)
+    {
+        // Cleanup is done in dissconnected callback.
+        mTcpTimer.Stop();
+        mEndpoint.SendEndOfStream();
+    }
+
+    VerifyOrExit(aBytesAvailable != 0);
+    mEndpoint.ReceiveByReference(data);
+    message = mSocket.NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    if (data != nullptr)
+    {
+        totalReceived += data->mLength;
+        message->AppendBytes(data->mData + kTcpQueryLenSize, data->mLength - kTcpQueryLenSize);
+        length = ReadUint16(data->mData);
+        VerifyOrExit(length != aBytesAvailable);
+
+        for (data = data->mNext; data != nullptr; data = data->mNext)
+        {
+            message->AppendBytes(data->mData, data->mLength);
+            totalReceived += data->mLength;
+        }
+    }
+
+    mEndpoint.CommitReceive(totalReceived, 0);
+    response.mInstance = &Get<Instance>();
+    response.mMessage  = message;
+
+    SuccessOrExit(ParseResponse(response, type, responseError));
+
+    if (responseError == kErrorNone && length > OPENTHREAD_CONFIG_DNS_OVER_TCP_QUERY_MAX_SIZE)
+    {
+        LogWarn("Dns query over wasn't received message too big.");
+        responseError = kErrorNoBufs;
+    }
+
+    FinalizeQuery(response, type, responseError);
+
+exit:
+    if (message != nullptr)
+    {
+        message->Free();
+    }
+    return;
+}
+
+void Client::HandleTcpReceiveAvailableCallback(otTcpEndpoint *aEndpoint,
+                                               size_t         aBytesAvailable,
+                                               bool           aEndOfStream,
+                                               size_t         aBytesRemaining)
+{
+    LogDebg("Dns tcp connection receive available callback.\n");
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))
+        ->HandleTcpReceiveAvailable(aEndpoint, aBytesAvailable, aEndOfStream, aBytesRemaining);
+}
+
+void Client::HandleTcpDisconnected(otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    Query    *query;
+    QueryInfo info;
+
+    LogDebg("Dns TCP disconnected reason: %d\n", aReason);
+    mEndpoint.Deinitialize();
+    mTcpInitialized = false;
+    VerifyOrExit(aReason != OT_TCP_DISCONNECTED_REASON_NORMAL);
+
+    // Abort queries in case of connection failures
+    for (Message &msg : mQueries)
+    {
+        query = static_cast<Query *>(&msg);
+        info.ReadFrom(*query);
+        if (info.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
+        {
+            FinalizeQuery(*query, kErrorAbort);
```

Comment:
One possible edge-case here:
If we want to allow TCP to used even if config say UDP (e.g. for larger message size or after some number of tx attempts) we need to handle it here as well (the config may say UDP transport but we may have actually used TCP for it).